File: 600/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(findIntegers(5))
}

func findIntegers(n int) int {
	ans := 0
	dp := [31]int{1,1}
	// fmt.Println(dp)
	for i := 2; i < 31; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}
	for i, pre := 29,0; i >= 0; i-- {
		val := 1 << i
		if n & val > 0 {
			ans += dp[i+1]
			if pre == 1 {
				break
			}
			pre = 1
		} else {
			pre = 0
		}
		if i == 0 {
			ans++
		}
	}
	return ans
}

File: 600/main.rs
#[warn(unused_comparisons)]

struct Solution {}


impl Solution {
    pub fn find_integers(n: i32) -> i32 {
        let mut ans = 0;
        let mut dp = vec![0; 31];
        dp[0] = 1;
        dp[1] = 1;
        // println!("{:#?}", dp);
        for i in 2..31 {
            dp[i] = dp[i-1] + dp[i-2]
        }
        let mut i: i32 = 29;
        let mut pre = 0;
        while i >= 0 {
            let val = 1 << i;
            if n & val > 0 {
                ans += dp[i as usize +1];
                if pre == 1 {
                    break;
                }
                pre = 1;
            } else {
                pre = 0;
            }
            if i == 0 {
                ans += 1;
            }
            i -= 1;
        }
        ans as i32
    }
}

fn main() {
    println!("{}", Solution::find_integers(5));
}